Make CamlinternalOO a bit less unsafe

chambart opened the pull request:
CamlinternalOO used to modify fields of blocks allocated as immutable,
which is unsafe and is going to become even more unsafe. An inline
record is used to allow marking a sum type as mutable.

This patch is extracted from the flambda branch

chambart committed:
Make CamlinternalOO a bit less unsafe

CamlinternalOO used to modify fields of blocks allocated as immutable,
which is unsafe and is going to become even more unsafe. An inline
record is used to allow marking a sum type as mutable.

alainfrisch commented:
This is a case where being able to refer to the inline record type would help.  Even without that, one could get rid of the unsafe "mut" function and "mut_table" type, at the price of one extra pattern matching in "lookup_key".  Do you have an idea whether this would affect performance in an observable way (my intuition is that it should be fine, considering that the function does allocations and mutations)?

chambart commented:
I didn't try to see the performance cost. I think the the cost would effectively be negligible. I kept it that way to keep the patch as simple as possible, but I will probably try that to see how far this would propagate.

chambart committed:
Get rid of some more casts in CamlinternalOO

Turn CamlinternalOO.tables into a GADT with cases distinguished by a
polymorphic variant. This allows to consider the Cons case almost like a
record. There is still one assumption that cannot be checked by typing:
when a key is found in the table, the data field is assumed to be a
Cons. This is dynamically checked. The cost is assumed to be
negligible (The branch should be well predicted by the processor).

chambart commented:
With a GADT I could make something that behave almost like if we could refer to the inline record. Yet we still need one 'mut' (the match and the assert false) since some external invariant is assumed.

alainfrisch commented:
I'm not 100% convinced by the complexity/benefit ratio of the last commit (compared to the first version + extra pattern matching to avoid "mut").

chambart committed:
Clean cruft from camlinternalOO

chambart committed:
Change static to dynamic checks in camlinternalOO

chambart commented:
I tried that (see the last patch).
I'm not convinced that this is clearer than the type annotated version.

mshinwell commented:
Alain, as far as I understand it, Pierre has tried your suggestion and the resulting code wasn't any clearer.
I'd like to merge this patch as soon as we can.  Are there any further objections?

gasche commented:
The fact that the patch uses inline records, that is a feature that is not available in any released version of OCaml, is giving me pause. I think it would be worth discussing that with other maintainers, and maybe having a specific mention to that effect in the changelog.

(I understand that doing this is the best solution, as flambda really needs this patch, and lower camlinternalOO performances may not be acceptable, but a discussion can help in eg. make sure people are aware of it and can find the change back if it bites them in some way in the future.)

xavierleroy commented:
Trying to better understand @gasche's reservation:

It's true that, so far, we've avoided using bleeding edge new features in the bootstrap path.  That path includes the bytecode compiler and the parts of the standard library that it uses.

It's also true that CamlintermalOO is very intimately tied to the compilation of objects and classes.

This said, AFAIK, the bootstrap path doesn't use objects and classes.  The native-code compiler does use them, but it is not part of the bootstrap path.

So, I'm not too scared by this patch, from a bootstrapping standpoint.

Is there another concern I overlooked?

alainfrisch commented:
> the bootstrap path doesn't use objects and classes

pprintast is linked in ocamlc and it creates a class.  I'm not overly concerned, though.  If needed, it would be easy to remove this dependency, and anyway, the fact that a release exists or not doesn't really change the story:  we can always go back to any previous version of the bootstrap compiler. (As for the stability of inline records, they have been used in production for about a year now.)

gasche commented:
@xavierleroy I think, aside the hard bootstrap issues, one justification for avoiding the bleeding edge is that there are various use cases for using tooling to work on the OCaml compiler codebase, and the ability to use tooling working with released versions of OCaml is a convenience in this situation. (One would think of Merlin, but Merlin has an experimental branch following trunk in any case, so this is more a general concern that a particular concern for a tool I use personally.) One could also think of situations where people would like to avoid depending on the bootstrap compiler (security issues?) but instead use an alternative implementation (js_of_ocaml, ocaml-java?) to compile the compiler sources.

That said, I think of it as a general rationale that is best followed when it is easy to do so. In the present situation, we really seem to need inline records to support flambda-correct code without performance degradation, so I'm all for it (having two good features that interact together in the same release is a sign of success, I guess).

damiendoligez commented:
The problem I see with using bleeding-edge features is that they are prone to incompatible changes, which would make our life harder. In the case of inline records, I don't think we're in great danger.

mshinwell commented:
So everyone is happy?  Xavier?

xavierleroy commented:
Fine with me.
